k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt

```diff
@@ -0,0 +1,99 @@
+import java.util.*
+
+class CredentialsResourceHandler(
+    cloudDriverK8sService: CloudDriverK8sService, taskLauncher: TaskLauncher, eventPublisher: EventPublisher,
+    orcaService: OrcaService, resolvers: List<Resolver<*>>
+) : GenericK8sResourceHandler<CredentialsResourceSpec, K8sObjectManifest>(
+    cloudDriverK8sService, taskLauncher,
+    eventPublisher, orcaService, resolvers
+) {
+    override val supportedKind = CREDENTIALS_RESOURCE_SPEC_V1
+    private val encoder: Base64.Encoder = Base64.getMimeEncoder()
+
+    public override suspend fun toResolvedType(resource: Resource<CredentialsResourceSpec>): K8sObjectManifest {
+        val cred = cloudDriverK8sService.getCredentialsDetails(resource.serviceAccount, resource.spec.account)
+        val data: MutableMap<String, String> = mutableMapOf()
+        // Priority: token > password > ssh key
+        when {
+            cred.token.isNotBlank() -> {
+                log.debug("populating token with username in manifest")
+                data.putAll(
+                    mapOf(
+                        "username" to encoder.encodeToString(cred.token.toByteArray()),
+                        "password" to ""
+                    )
+                )
+            }
+            cred.username.isNotBlank() -> {
+                log.debug("populating username and password in manifest")
+                data.putAll(
+                    mapOf(
+                        "username" to encoder.encodeToString(cred.username.toByteArray()),
+                        "password" to encoder.encodeToString(cred.password.toByteArray())
+                    )
+                )
+            }
+            cred.sshPrivateKey.isNotBlank() -> {
+                log.debug("populating private key in manifest")
+                data.putAll(mapOf("identity" to encoder.encodeToString(cred.sshPrivateKey.toByteArray())))
+            }
+        }
+        // setting strategy.spinnaker.io/versioned is needed to avoid creating secrets with versioned names e.g. testing1-v000
+        return K8sObjectManifest(
+            "v1",
+            "Secret",
+            mapOf(
+                "namespace" to resource.spec.namespace,
+                "name" to resource.spec.account,
+                "annotations" to mapOf("strategy.spinnaker.io/versioned" to "false")
+            ),
+            null,
+            data as MutableMap<String, Any>?
+        )
+    }
+
+    override suspend fun current(resource: Resource<CredentialsResourceSpec>): K8sObjectManifest? {
+        log.debug("resource received: $resource spec: ${resource.spec}")
+        try {
+            val res = cloudDriverK8sService.getK8sResource(
+                resource.serviceAccount,
+                resource.spec.locations.account,
+                resource.spec.namespace,
+                "secret ${resource.spec.account}"
+            )
+            log.debug("response from clouddriver: $res manifest: ${res.manifest}")
+            return res.manifest
```

Comment:
I will have to still do some thorough testing, but isn't it that the `data` field for a k8s secret is `base64` encoded when retrieved from the cluster? i will have to run it against the cluster, but if so, then is there a diff always between your `current` and `desired` objects, since one is a map and another one is base64 encoded data?